crates/noirc_evaluator/src/ssa/optim.rs

```diff
@@ -1,225 +1,151 @@
-            }
-            obj
-        }
-    }
-}
 
 // Performs constant folding, arithmetic simplifications and move to standard form
-pub fn simplify(ctx: &mut SsaContext, ins: &mut node::Instruction) {
+// Returns a new NodeId to replace with the current if the current is deleted.
+pub fn simplify(ctx: &mut SsaContext, ins: &mut node::Instruction) -> Option<NodeId> {
     //1. constant folding
-    let l_eval = to_const(ctx, NodeEval::VarOrInstruction(ins.lhs));
-    let r_eval = to_const(ctx, NodeEval::VarOrInstruction(ins.rhs));
-    let idx = match ins.evaluate(&l_eval, &r_eval) {
+    let new_id = match ins.evaluate(ctx) {
         NodeEval::Const(c, t) => ctx.get_or_create_const(c, t),
         NodeEval::VarOrInstruction(i) => i,
     };
-    if idx != ins.id {
-        ins.is_deleted = true;
-        ins.rhs = idx;
-        if idx == NodeId::dummy() {
-            ins.operator = node::Operation::Nop;
+
+    if new_id != ins.id {
+        ins.replacement = Some(new_id);
+        if new_id == NodeId::dummy() {
+            ins.delete();
         }
-        return;
+        return Some(new_id);
     }
 
     //2. standard form
     ins.standard_form();
     match ins.operator {
-        node::Operation::Cast => {
-            if let Some(lhs_obj) = ctx.try_get_node(ins.lhs) {
-                if lhs_obj.get_type() == ins.res_type {
-                    ins.is_deleted = true;
-                    ins.rhs = ins.lhs;
-                    return;
+        Operation::Cast(value_id) => {
+            if let Some(value) = ctx.try_get_node(value_id) {
+                if value.get_type() == ins.res_type {
+                    ins.delete();
+                    return Some(NodeId::dummy());
                 }
             }
         }
-        // node::Operation::Gte => {
-        //     //a>=b <=> Not(a<b)
-        //     let inv = eval.new_instruction(ins.lhs, ins.rhs, node::Operation::Lt, ins.res_type);
-        //     ins.lhs = eval.get_const(FieldElement::one(), ins.res_type);
-        //     ins.rhs = inv;
-        //     ins.operator = node::Operation::Sub; //n.b. no need to underflow here
-        //TODO: inv must be inserted before ins.
-        // }
-        node::Operation::Constrain(op) => match op {
-            node::ConstrainOp::Eq => {
-                if let (Some(a), Some(b)) =
-                    (mem::Memory::deref(ctx, ins.lhs), mem::Memory::deref(ctx, ins.rhs))
-                {
-                    if a == b {
-                        ins.is_deleted = true;
-                        ins.operator = node::Operation::Nop;
-                    }
+        Operation::Binary(node::Binary { operator: BinaryOp::Constrain(op), lhs, rhs }) => {
+            match (op, Memory::deref(ctx, lhs), Memory::deref(ctx, rhs)) {
+                (ConstrainOp::Eq, Some(lhs), Some(rhs)) if lhs == rhs => {
+                    ins.delete();
                 }
+                _ => (),
             }
-            node::ConstrainOp::Neq => {
```

Comment:
It only contained an assert for something that it seemed to me we did not want to panic at runtime for, just fail during verification.

Reply:
No it should be here, may be we could add a better error message later on.

Reply:
I agree to discuss this around an issue. But since this PR is about the refactoring of the code, you should keep the behavior and add the assert back. I prefer to crash the compiler rather than producing a fake proof.

Reply:
This seems to be blocking, so lets leave this as is since there is an issue for it. Proofs that fail to verify are not fake proofs, they're valid use cases. Just ones that we want better errors/warnings for when possible.

Reply:
See my other comment; failing proofs are not a valid use case, so let's leave the assert and solve this through the new issue.

Reply:
re-added